### PR TITLE
Add warning message when leaving Dashboard Edit Page

### LINKED ISSUE #29971: FE - Editing a dashboard

### PR DESCRIPTION
**Pull request by oisincoveney**
Closes #29971 

### Description

Adds the browser's "Are you sure you want to leave" page when the user is editing a dashboard. This uses a pure JS event listener for `onBeforeUnload` to prompt the user using the native browser popup when the user tries to navigate away from the page (i.e. tries to close the tab, go to a new URL, or refresh the page).

### How to verify

Describe the steps to verify that the changes are working as expected.

1. Go to any dashboard and start to edit.
2. Try to refresh the page, and the browser should prompt the user with a "Are you sure you want to leave" similar to the below screenshot - the message might be different in your specific browser.

### Demo

image

### Checklist

- [ ] Tests have been added/updated to cover changes in this PR

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/29986)

### CODE REVIEW

File: frontend/src/metabase/lib/dom.js [hunk 1]
```diff
@@ -513,3 +513,9 @@ export const getEventTarget = event => {
 
   return target;
 };
```
Comment (oisincoveney) [inline frontend/src/metabase/lib/dom.js:515]: Added this here for now, since I already know I want to reuse this in the other tasks within the epic

File: frontend/src/metabase/dashboard/components/Dashboard/Dashboard.jsx [hunk 2]
```diff
@@ -129,6 +140,8 @@ class Dashboard extends Component {
     main.addEventListener("resize", this.throttleParameterWidgetStickiness, {
       passive: true,
     });
+
```
Comment (oisincoveney) [inline frontend/src/metabase/dashboard/components/Dashboard/Dashboard.jsx:143]: Would there be a better way to do this? I feel like writing `window.addEventListener` in a bunch of components might seem a little copy-pastey for my liking, not sure if we can have a hook or something that we can use across the class and functional components that the components in this epic are going to use

Comment (ranquild) [reply frontend/src/metabase/dashboard/components/Dashboard/Dashboard.jsx:143]: We could use `useBeforeUnload` hook from `react-use` in `DashboardApp` component, which is a functional component.

File: frontend/src/metabase/dashboard/containers/DashboardApp.jsx [hunk 3]
```diff
@@ -132,6 +132,10 @@ const DashboardApp = props => {
 
   useUnmount(props.reset);
 
+  // most browsers don't use a custom message anymore, just putting here to retain compatibility
+  // https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event#compatibility_notes
+  useBeforeUnload(isEditing && isDirty, "You have unsaved changes.");
+
```
Comment (kulyk) [inline frontend/src/metabase/dashboard/containers/DashboardApp.jsx:138]: Please wrap the string into `t` from `ttag` to make sure it can be translated:

```js
import { t } from "ttag";

onBeforeOnload(isEditing && isDirty, t`You have unsaved changes.`);
```

Comment (oisincoveney) [reply frontend/src/metabase/dashboard/containers/DashboardApp.jsx]: Done 😄

File: e2e/test/scenarios/dashboard/dashboard.cy.spec.js [hunk 4]
```diff
@@ -178,6 +182,18 @@ describe("scenarios > dashboard", () => {
     cy.findByText("Orders, Count");
   });
 
+  it("should trigger window.onbeforeunload when trying to leave an unsaved, edited dashboard", () => {
```
Comment (ranquild) [inline e2e/test/scenarios/dashboard/dashboard.cy.spec.js:185]: Not sure how to test this but the current doesn't fail if we remove the change in the code. Could we improve the test so it fails if there is no message?

File: frontend/src/metabase/dashboard/containers/DashboardApp.jsx [hunk 5]
```diff
@@ -132,6 +132,10 @@ const DashboardApp = props => {
 
   useUnmount(props.reset);
 
+  // most browsers don't use a custom message anymore, just putting here to retain compatibility
```
[the remark this answers is not in the document]
Comment (oisincoveney) [reply frontend/src/metabase/dashboard/containers/DashboardApp.jsx:135]: I think I have a compromise - I'll hoist the string to a constant that we can use across all of the future `useBeforeUnload` calls, and leave the comment with that constant. I think it's worth having the comment to the MDN docs there for some explicit context, but I agree we shouldn't have it everywhere

File: frontend/src/metabase/dashboard/containers/DashboardHeader.jsx [hunk 6]
```diff
@@ -97,7 +97,7 @@ class DashboardHeader extends Component {
     onChangeLocation: PropTypes.func.isRequired,
 
     toggleSidebar: PropTypes.func.isRequired,
-    sidebar: PropTypes.string.isRequired,
+    sidebar: PropTypes.object.isRequired,
```
Comment (oisincoveney) [inline frontend/src/metabase/dashboard/containers/DashboardHeader.jsx:100]: From my testing it seems like `sidebar` always takes in:

```
{
    props: {...},
    name: <sidebar type name>,
}
```

File: frontend/src/metabase/dashboard/constants.js [hunk 7]
```diff
@@ -6,3 +8,7 @@ export const SIDEBAR_NAME = {
   sharing: "sharing",
   info: "info",
 };
+
+// most browsers don't use a custom message with beforeunload anymore, just putting here to retain compatibility
```
Comment (oisincoveney) [inline frontend/src/metabase/dashboard/constants.js:12]: I put this constant here but I'm anticipating that it will be used across the other tasks. Where else can I put this? I was thinking somewhere in `core` but not totally sure

File: frontend/src/metabase/dashboard/containers/DashboardApp.jsx [hunk 8]
```diff
@@ -33,29 +33,30 @@ import * as Urls from "metabase/lib/urls";
...
-  getIsEditingParameter,
-  getIsDirty,
   getDashboardComplete,
-  getCardData,
-  getSlowCards,
+  getDocumentTitle,
   getEditingParameter,
+  getFavicon,
+  getIsAdditionalInfoVisible,
+  getIsAddParameterPopoverOpen,
```
Comment (oisincoveney) [inline frontend/src/metabase/dashboard/containers/DashboardApp.jsx:47]: This was Prettier's doing

File: frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+function createMockEventListener() {
+  const events = {};
+  const mockEvent = {
+    preventDefault: jest.fn(),
+    returnValue: undefined,
+  };
+
+  window.addEventListener = jest.fn((event, callback) => {
```
Comment (ranquild) [inline frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx:73]: It seems that we don't revert this change. Maybe `jest.spyOn(window, "addEventListener")` and then `jest.restoreAllMocks();` in `afterEach` would do the trick?

Comment (oisincoveney) [reply frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx:73]: yep, makes sense and will help with consistency with the other code

File: frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+});
+
+async function setup(user = createMockUser()) {
+  setupDashboardEndpoints(createMockDashboard(mockDashboard));
+  setupCollectionsEndpoints([]);
+  setupCollectionItemsEndpoint();
+
+  fetchMock.get("path:/api/bookmark", []);
+
+  jest.spyOn(document, "querySelector").mockImplementation(() => ({
```
[the remark this answers is not in the document]
Comment (ranquild) [reply frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx:38]: Could we render this link in our `setup` function instead? Like `<link rel="icon" />`

File: frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+    setAttribute: jest.fn(),
+    classList: {
+      remove: jest.fn(),
+    },
+  }));
+
+  jest
```
Comment (ranquild) [inline frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx:45]: Same for this one - I'd rather fix our test suite setup

File: frontend/test/__support__/server-mocks/collection.ts [hunk 12]
```diff
@@ -58,3 +58,19 @@ export function setupCollectionVirtualSchemaEndpoints(
   fetchMock.get(urls.questions, questionVirtualTables);
   fetchMock.get(urls.models, modelVirtualTables);
 }
+
+export function setupCollectionItemsEndpoint(
+  collectionItems: CollectionItem[] = [],
+  models = ["card"],
+  limit = collectionItems.length,
+  offset = 0,
+  opts: Record<string, unknown> = {},
```
[the remark this answers is not in the document]
Comment (ranquild) [reply frontend/test/__support__/server-mocks/collection.ts:65]: Yes looks good. Could you adjust the following:
- Accept `collectionItems` via the argument and no construct them here
- Remove `opts` - seems not used
- use `collection.id` in the endpoint url, e.g. `fetchMock.get("path:/api/collection/${collection.id}/items")`

File: frontend/src/metabase/dashboard/components/DashboardHeader.tsx [hunk 13]
```diff
@@ -90,7 +90,11 @@ const DashboardHeader = ({
         className="Header-buttonSection"
         isNavBarOpen={isNavBarOpen}
       >
-        {headerButtons}
+        {headerButtons.map((headerButton, index) => (
```
Comment (ranquild) [inline frontend/src/metabase/dashboard/components/DashboardHeader.tsx:93]: Could you explain this change?

File: frontend/src/metabase/dashboard/components/DashboardHeader.tsx [hunk 14]
```diff
@@ -90,7 +90,11 @@ const DashboardHeader = ({
...
+          <div style={{ display: "contents" }} key={index}>
```
Comment (uladzimirdev) [inline frontend/src/metabase/dashboard/components/DashboardHeader.tsx:94]: why not to use import from "styled"?

File: frontend/src/metabase/dashboard/constants.js [hunk 15]
```diff
@@ -6,3 +8,7 @@ export const SIDEBAR_NAME = {
...
+// https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event#compatibility_notes
+export const BEFORE_UNLOAD_UNSAVED_MESSAGE = t`You have unsaved changes.`;
```
Comment (ranquild) [inline frontend/src/metabase/dashboard/constants.js:14]: Maybe we should add a hook to `metabase/core/hooks` that wraps `useBeforeUnload` with this message?

File: frontend/src/metabase/dashboard/constants.js (hunk 7 above)
Comment (uladzimirdev) [reply frontend/src/metabase/dashboard/constants.js:12]: I'd use it close to the code usage maybe? if many places are involved, then it's fine

File: frontend/src/metabase/dashboard/containers/DashboardHeader.jsx (hunk 6 above)
Comment (uladzimirdev) [reply frontend/src/metabase/dashboard/containers/DashboardHeader.jsx:100]: I have a better fix for it 😄 but great you touched it

File: frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (uladzimirdev) [reply frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx:38]: I know it's a bit hacky, but what if you use something like this? another idea is just to mock `withFavicon` HOC

```diff
+document.querySelector('link[rel="icon"]')?.setAttribute("href", favicon);
-document.querySelector('link[rel="icon"]').setAttribute("href", favicon);
```

File: frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+  });
+
+  it("should have a beforeunload event when the user tries to leave a dirty dashboard", async function () {
+    const { events, mockEvent } = await setup();
+
+    userEvent.click(screen.getByLabelText("Edit dashboard"));
+    userEvent.click(screen.getByTestId("dashboard-name-heading"));
+    userEvent.type(screen.getByTestId("dashboard-name-heading"), "a");
+    // need to click away from the input to trigger the isDirty flag
+    userEvent.click(screen.getByLabelText("Add questions"));
```
Comment (uladzimirdev) [inline frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx:104]: probably `userEvent.tab()` would be simpler?

**Review by uladzimirdev — COMMENTED** [review]
You did a great job here, but I'm concerned about the unit tests, they are pretty complicated and *maybe* don't make a lot of sense if you managed to find a solution for e2e test case

File: frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx (hunk 16 above)
Comment (oisincoveney) [reply frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx:104]: Ah definitely. Good call

File: frontend/src/metabase/dashboard/components/DashboardHeader.tsx (hunk 13 above)
Comment (oisincoveney) [reply frontend/src/metabase/dashboard/components/DashboardHeader.tsx:93]: I responded to @uladzimirdev with this, what do you think?

> Ah ignore this. my local tests weren't passing because of the whole each child must have a key when it was just {headerButtons} so I did this as a bandaid that I would take out. Do you think it would make more sense to keep this here and move it to a styled component? Then it would at least get rid of one of the errors

File: frontend/src/metabase/dashboard/containers/DashboardHeader.jsx (hunk 6 above)
Comment (oisincoveney) [reply frontend/src/metabase/dashboard/containers/DashboardHeader.jsx:100]: Oooh ok - I'd much rather use a fix from someone who knows their way around a little better, what were you planning to use here?

File: frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (oisincoveney) [reply frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx:38]: Yeah I was thinking about just adding that optional chaining, I just didn't want to modify any code that wasn't directly related to the test that I was making - I'll have a look at @ranquild 's solution first but if it doesn't work we can think about using this one

File: frontend/src/metabase/dashboard/constants.js (hunk 7 above)
Comment (oisincoveney) [reply frontend/src/metabase/dashboard/constants.js:12]: Yep, I've moved it into a custom hook that will keep the message in one place with the `react-use` hook, so we can still import the message for tests and then reuse the code without having to import the message everywhere in the code

File: frontend/src/metabase/dashboard/components/DashboardHeader.tsx (hunk 14 above)
Comment (iethree) [reply frontend/src/metabase/dashboard/components/DashboardHeader.tsx:94]: 1) you shouldn't ever use indexes for keys in react because it can cause some insane and really hard to track down bugs if the order of the items ever changes
2) if some of these items are missing keys, I think it would be better to add them wherever `headerbuttons` is coming from

Comment (oisincoveney) [reply frontend/src/metabase/dashboard/components/DashboardHeader.tsx:94]: Ah yeah, it's a horrible practice - I basically did this just to make sure I wouldn't see that error. I was trying to add the keys to the header buttons, but I was seeing a few `Fragment` components so I didn't want to spend time adding the keys when the code was already working fine before, especially when changing that code was also not really in the scope of this PR

Comment (oisincoveney) [reply frontend/src/metabase/dashboard/components/DashboardHeader.tsx:94]: From what I see, the options seem to be:
- leave it and add the keys and things later in a different PR
- Make the changes now
I just don't want to muddle the PR - let me know what you tink

File: frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx (hunk 16 above)
Comment (oisincoveney) [reply frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx:106]: Hm... this seemed to bork when I tried to change this to `userEvent.tab` (something about being unable to find a different place to tab to?). I'll revert this to the original thing I had and look at it tomorrow

Comment (oisincoveney) [reply frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx:105]: Ah ok it was the same problem with my node version and the fact that the dom couldn't be found for some reason. Moving to node 18 solved my issue

File: frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (oisincoveney) [reply frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx]: Solved adding the favicon directly to the document head, and wrapping the `DashboardApp` component in a `<main>`

File: frontend/src/metabase/dashboard/components/DashboardHeader.tsx (hunk 14 above)
Comment (oisincoveney) [reply frontend/src/metabase/dashboard/components/DashboardHeader.tsx]: Ok - seems like it was just adding one key to the `Fragment` for the `Add actions` button, so I just added it and seems to be fine

File: frontend/src/metabase/hooks/use-before-unload/use-before-unload.ts [hunk 17]
```diff
@@ -0,0 +1,12 @@
+import { useBeforeUnload as useBeforeUnloadHook } from "react-use";
+import { t } from "ttag";
+
+// most browsers don't use a custom message with beforeunload anymore, just putting here to retain compatibility
+// https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event#compatibility_notes
+export const BEFORE_UNLOAD_UNSAVED_MESSAGE = t`You have unsaved changes.`;
+
+const useBeforeUnload = (condition: (() => boolean) | boolean) => {
```
Comment (uladzimirdev) [inline frontend/src/metabase/hooks/use-before-unload/use-before-unload.ts:8]: nit: it can be too much, but consider this 😄 

```diff
-const useBeforeUnload = (condition: (() => boolean) | boolean) => {
+type Condition = Parameters<typeof useBeforeUnloadHook>[0]
+const useBeforeUnload = (condition: Condition) => {
   return useBeforeUnloadHook(condition, BEFORE_UNLOAD_UNSAVED_MESSAGE);
 };
 ```

**Review by uladzimirdev — COMMENTED** [review]
To me it looks good

File: frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+
+  await waitForElementToBeRemoved(() =>
+    screen.queryAllByTestId("loading-spinner"),
+  );
+
+  return { mockEvent, events };
+}
+
+function createMockEventListener() {
+  const events = {};
```
Comment (ranquild) [inline frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx:68]: Maybe we could simplify the setup.

```
const setup = (...) => {
  const eventMock = { preventDefault: jest.fn() });
  const eventListenerMock = jest.spyOn(window, "addEventListener");

  return { eventMock, eventListenerMock };
});
```

and the test:

```
const { eventMock, eventListenerMock } = await setup(); 

...

eventListenerMock.mock.calls
  .filter(([event]) => event === "beforeunload"))
  .forEach(([_, callback]) => callback(mockEvent));

expect(eventMock.preventDefault).toHaveBeenCalled();
expect(eventMock.returnValue).toBe(...);
```

**Review by ranquild — COMMENTED** [review]
Looks good, I've left a suggestion how we can make the setup easier and also move the mocking logic to `setup` function

File: frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx (hunk 18 above)
Comment (oisincoveney) [reply frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx:68]: It looks good, I think the only thing I'm concerned about is reusability - the reason I put it in this function in the first place was with the thinking that we could use this in other files

File: frontend/src/metabase/hooks/use-before-unload/use-before-unload.ts (hunk 17 above)
Comment (oisincoveney) [reply frontend/src/metabase/hooks/use-before-unload/use-before-unload.ts:8]: More typescript is (almost) always good 😄

File: frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+  setupCollectionsEndpoints([]);
+  setupCollectionItemsEndpoint(TEST_COLLECTION);
+
+  fetchMock.get("path:/api/bookmark", []);
+
+  const mockEventListener = jest.spyOn(window, "addEventListener");
+  const mockEvent = {
+    preventDefault: jest.fn(),
+  };
+  const callMockEvent = eventNames => {
```
Comment (oisincoveney) [inline frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx:42]: Maybe this is overkill, let me know

File: frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+  createMockDashboard,
+  createMockUser,
+} from "metabase-types/api/mocks";
+import {
+  setupCollectionItemsEndpoint,
+  setupCollectionsEndpoints,
+  setupDashboardEndpoints,
+} from "__support__/server-mocks";
+import { createMockDashboardState } from "metabase-types/store/mocks";
+import { BEFORE_UNLOAD_UNSAVED_MESSAGE } from "metabase/hooks/use-before-unload";
```
Comment (kulyk) [inline frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx:20]: 1. You can import everything you can find in `@testing-library/react` from `__support__/ui`; could we please keep only one import?
2. Not sure why linter doesn't complaint, but `metabase/hooks` import should come _before_ `metabase-types` imports

File: frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+  await waitForElementToBeRemoved(() =>
+    screen.queryAllByTestId("loading-spinner"),
+  );
+
+  return { mockEvent, mockEventListener, callMockEvent };
+}
+
+describe("DashboardApp", function () {
+  beforeAll(() => {
+    const linkElem = document.createElement("link");
```
Comment (ranquild) [inline frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx:77]: nit: would it be possible to render this link inside `DashboardAppContainer`?

File: frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+
+  const mockEventListener = jest.spyOn(window, "addEventListener");
+  const mockEvent = {
+    preventDefault: jest.fn(),
+  };
+  const callMockEvent = eventNames => {
+    mockEventListener.mock.calls
+      .filter(([event]) => eventNames.includes(event))
+      .forEach(([_, callback]) => callback(mockEvent));
+  };
```
Comment (kulyk) [inline frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx:46]: Could you please leave a comment explaining what exactly are we trying to mock here?

File: frontend/src/metabase/dashboard/containers/DashboardHeader.jsx [hunk 23]
```diff
@@ -228,11 +228,12 @@ class DashboardHeader extends Component {
...
       buttons.push(
-        <Tooltip tooltip={addQuestionButtonHint}>
+        <Tooltip key="add-question-element" tooltip={addQuestionButtonHint}>
           <DashboardHeaderButton
             icon="add"
             isActive={activeSidebarName === SIDEBAR_NAME.addQuestion}
             onClick={() => toggleSidebar(SIDEBAR_NAME.addQuestion)}
+            aria-label={t`Add questions`}
             data-metabase-event="Dashboard;Add Card Sidebar"
           />
```
Comment (kulyk) [inline frontend/src/metabase/dashboard/containers/DashboardHeader.jsx:238]: Thank you ❤️

File: frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx (hunk 19 above)
Comment (ranquild) [reply frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx:42]: nit: maybe we actually only need to setup `mockEventListener` here and then have a separate function for the call like:
```
const callMockEvent = (mockEventListener, eventName) => {
  const mockEvent = { preventDefault: jest.fn() };

  mockEventListener.mock.calls
      .filter(([event]) => eventNames.includes(event))
      .forEach(([_, callback]) => callback(mockEvent));

  return mockEvent;
};
```

**Review by ranquild — APPROVED** [review]

**Review by kulyk — APPROVED** [review]

File: frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx (hunk 20 above)
Comment (oisincoveney) [reply frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx:20]: Hmm yeah the linter doesn't seem to care where I put the `metabase/hooks` import as long as it isn't with the 3rd party library imports. Not sure what happened there, I checked my eslint config and it's pointing to our `eslintrc` file

File: frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx (hunk 21 above)
Comment (oisincoveney) [reply frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx:77]: Weirdly enough, we tried that yesterday and it didn't work. Now I tried it again and it works perfectly...

File: frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx [hunk 24]
```diff
@@ -0,0 +1,110 @@
...
+const TEST_DASHBOARD = createMockDashboard({
+  id: 1,
+  name: "Example",
+});
+
+const TEST_COLLECTION = createMockCollection({
+  id: "root",
+});
+
+async function setup(user = createMockUser()) {
```
[the remark this answers is not in the document]
Comment (oisincoveney) [reply frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx:31]: I think it's a good idea  - more flexibility is (almost) always a good thing 😄

**Review by mazameli — COMMENTED** [review]
Worked for me in Arc on refreshing or closing the tab, but I was expecting it to work if I click the browser's back button, and it did not. Is it supposed to?

**Comment by oisincoveney** [comment]
> Worked for me in Arc on refreshing or closing the tab, but I was expecting it to work if I click the browser's back button, and it did not. Is it supposed to?

Not for now - from what I've read (https://stackoverflow.com/questions/40817395/react-js-detect-backbutton for the most succinct explanation), the back button handles things differently when we're in the app due to React Router and the fact that it's a SPA. So _technically_ when using the back button, the browser doesn't recognize that we're leaving the page since we're staying in the app.

**Comment by mazameli** [comment]
@oisincoveney Got it, thanks. That's a bummer, and I think it would be rad if we could solve for the Back case, too — but I would not block shipping this on that, as the rest works well and adds value.

**Comment by oisincoveney** [comment]
@mazameli Gotcha, I'll ship this and then create a new issue for the back buttons. It's definitely a nice-to-have, just super weird to me that the browser handles that differently from the other "navigate away" events
